src/psyclone/psyir/transformations/acc_update_trans.py

```diff
@@ -0,0 +1,403 @@
+                    self._add_update_directives([child.start_expr,
+                                                 child.stop_expr,
+                                                 child.step_expr])
+                    # Recurse down into the loop body
+                    self._add_updates_to_schedule(child.loop_body)
+
+        # We've reached the end of the list of children - are there any
+        # last nodes that represent computation on the host?
+        self._add_update_directives(node_list)
+
+    def _add_update_directives(self, node_list):
+        '''
+        Adds the required update directives before and after the nodes in
+        the supplied list.
+
+        :param sched: the schedule which contains the nodes in node_list.
+        :type sched: :py:class:`psyclone.psyir.nodes.Schedule`
+        :param node_list: the PSyIR nodes representing code executed on the \
+                          host rather than the device.
+        :type node_list: list of :py:class:`psyclone.psyir.nodes.Node`
+
+        '''
+        if not node_list:
+            return
+
+        # TODO #1872: the lack of precise array access descriptions might
+        # unnecessarily increase the data transfer volume.
+        inputs, outputs = self._dep_tools.get_in_out_parameters(node_list)
+        inputs, outputs = set(inputs), set(outputs)
+
+        # TODO #1872: as a workaround for the lack of precise array access
+        # descriptions, we currently overapproximate dependencies by adding any
+        # variables being written to the input set (in addition to the output
+        # set). The aim is to avoid situations where the host only writes an
+        # array slice and the respective update device directive would then
+        # incorrectly overwrite the entire array on the device, including the
+        # data outside the slice. If or when this is no longer needed, the step
+        # of the algorithm below dealing with if stmts will need updating.
+        inputs.update(outputs)
+
+        # TODO #1732: for CodeBlock nodes, until we have semantic information
+        # on the statements involved, we conservatively assume every referenced
+        # variable is both an input and an output.
+        # TODO #1872: CodeBlock.get_symbol_names might include function names
+        # mistaken for array names. However, as these do not correspond to any
+        # arrays, the compiler should be able to ignore them in the directive.
+        for node in node_list:
+            if isinstance(node, CodeBlock):
+                for symbol_name in node.get_symbol_names():
+                    inputs.add(Signature(symbol_name))
+                    outputs.add(Signature(symbol_name))
+
+        # Copy any data that is read by this region to the host (resp. device)
+        # if it is on the device (resp. host).
+        IN, OUT = 0, 1
+        for idx, host_sig in enumerate((inputs, outputs)):
```

Comment:
The comments have helped a lot but I still find this code pretty hard to understand and for future maintenance that's an important consideration. Please could you therefore try replacing this loop with, e.g. two successive calls to a new, internal method, once for inputs and once for outputs. That way, for instance, `idx` won't get carried through the code and used to make decisions here and there (hopefully).
I'm not totally certain of the best way to refactor - you'll have to try and it and see what works best and/or we can discuss it together.

Reply:
I have refactored the code by creating a new internal method that finds the variables requiring synchronisation against a set of statements including compute regions. This allowed me to reduce the nesting depth to no more than three... Let me know what you think.